### PR TITLE
Line filters: Regex support

### PR DESCRIPTION
**Pull request by gtk-grafana**
Adds regex and exclusion support to the line filters.

Fixes: https://github.com/grafana/explore-logs/issues/742
Fixes: https://github.com/grafana/explore-logs/issues/421
Addresses: https://github.com/grafana/explore-logs/issues/758

> [!NOTE] 
> Breaking UX change:
> Line filters must be submitted before they will run queries.

This PR:
* adds exclusion operations (does not contain)
* regex operations (does|does not match)
* provides support for case sensitivity to work with regex/exclusion
* Upgrades scenes to 5.36
* allows multiple line filters

### Multiple line filters
image

### Regex line filters
image

### Exclusion filters from log line context menu
image
image

#### TODOs 
- [x] UI finalization: Line filter component
- [x] UI finalization: Ad hoc variables
- [x] Duplicate queries when submitting a filter (https://github.com/grafana/scenes/pull/1004)
- [x] Fix existing tests
- [x] Add test coverage for regex and negative matching
- [x] Discuss serializing case sensitivity option
- [x] Escape values
- [x] Add negative line filter to logs panel text selection
- [x] Migrate old line filters in URLs to new variables.

### Notes for your reviewers:
Part of https://github.com/grafana/explore-logs/issues/952

https://github.com/grafana/explore-logs/pull/958 wasn't going to work, see https://github.com/grafana/explore-logs/pull/958#issuecomment-2541865172 for more detail.

### CODE REVIEW

**Comment by gtk-grafana** [comment]
Found a bug where submitting after clearing adds an empty variable. Will fix and self-review after the holiday, but this is really close!!!!

**Comment by gtk-grafana** [comment]
We discussed this in standup today and the squad was leaning towards updating the current implementation to stop the query from executing on change, and either add a dedicated editing mode (add a new button which will open a modal with the line filter input and a logs panel limited to 100 lines), or to add an additional "preview" button which will execute the line filter without adding it to the global variables.

@zizzpudding We're going to hold-off on making more changes until we can meet and determine a path forward as the squad was split on the desired functionality of this component.

File: src/Components/ServiceScene/LineFilter/LineFilterScene.tsx [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+  /**
+   * Submits on enter
+   */
+  handleEnter = (e: KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'Enter') {
+      if (this.state.lineFilter) {
+        this.onSubmitLineFilter();
+      }
+    }
```
Comment (matyax) [inline src/Components/ServiceScene/LineFilter/LineFilterScene.tsx:188]:
```suggestion
    if (e.key === 'Enter' && this.state.lineFilter) {
      this.onSubmitLineFilter();
    }
```

File: src/Components/ServiceScene/LogsPanelScene.tsx [hunk 2]
```diff
@@ -139,6 +139,7 @@ export class LogsPanelScene extends SceneObjectBase<LogsPanelSceneState> {
         .setOption('onClickFilterOutLabel', this.handleLabelFilterOutClick)
         .setOption('isFilterLabelActive', this.handleIsFilterLabelActive)
         .setOption('onClickFilterString', this.handleFilterStringClick)
+        .setOption('onClickFilterOutString', this.handleFilterOutStringClick)
```
Comment (matyax) [inline src/Components/ServiceScene/LogsPanelScene.tsx:142]: Nice!

File: src/Components/ServiceScene/LogsPanelScene.tsx [hunk 3]
```diff
@@ -246,10 +248,35 @@ export class LogsPanelScene extends SceneObjectBase<LogsPanelSceneState> {
...
+          selectionLength: value.length,
+        }
+      );
+    }
+  };
+
   private handleFilterStringClick = (value: string) => {
     const parentModel = sceneGraph.getAncestor(this, LogsListScene);
     const lineFilterScene = parentModel.getLineFilterScene();
     if (lineFilterScene) {
```
Comment (matyax) [inline src/Components/ServiceScene/LogsPanelScene.tsx:274]: Remember to add it as an applied filter, and not populate the input instead. For filter and filter out.

**Review by matyax — COMMENTED** [review]
Amazing feature. Left some minor comments.

File: src/Components/ServiceScene/LineFilter/LineFilterScene.test.tsx [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        }),
+      });
+    });
+
+    test('Updates the variable with the user input', async () => {
+      render(<scene.Component model={scene} />);
+
+      const string = `((25[0-5]|(2[0-4]|1\\d|[1-9]|)\\d)\\.?\\b){4}`;
+      const input = screen.getByPlaceholderText('Search in log lines');
+      // Jest can't type regex apparently
```
[the remark this answers is not in the document]
Comment (gtk-grafana) [reply src/Components/ServiceScene/LineFilter/LineFilterScene.test.tsx:166]: Not really, but jest does have a bug where `userEvent.type` won't work with complex strings like this because the brackets and other chars are used for templating. I'll pull the comments and hopefully folks won't spend time bashing their head against this in the future

File: src/Components/ServiceScene/LogsPanelScene.tsx (hunk 3 above)
Comment (gtk-grafana) [reply src/Components/ServiceScene/LogsPanelScene.tsx]: Fixed in https://github.com/grafana/explore-logs/pull/963/commits/c1d9c5e8e227b1ec451c69952543e8daf1256781

**Comment by gtk-grafana** [comment]
Updated to no longer run queries on change, we can follow up with improvements to the debugging UX later.

**Comment by gtk-grafana** [comment]
Met with @zizzpudding we agreed to kick the debug/editing mode to a follow-up PR, and we tweaked the style of the submit button(s).

Disabled:
image

Enabled:
image

**Comment by gtk-grafana** [comment]
I would say this is ready for final review. Be nice to get this out in 1.0.5, but we can discuss tomorrow

**Comment by matyax** [comment]
Hard agree with the include/exclude, specially after I saw it in patterns too. I'll take another look in the am. Great work!

File: src/Components/ServiceScene/LogsListScene.tsx [hunk 5]
```diff
@@ -175,6 +172,24 @@ export class LogsListScene extends SceneObjectBase<LogsListSceneState> {
...
+
+    // Subscribe to line filter state so we can pass the current filter between different viz
+    if (this.state.panel) {
+      const lineFilterScenes = sceneGraph.findDescendents(this.state.panel, LineFilterScene);
+      if (lineFilterScenes.length) {
+        const lineFilterScene = lineFilterScenes[0];
+        this._subs.add(
+          lineFilterScene.subscribeToState((newState, prevState) => {
+            if (newState.lineFilter !== prevState.lineFilter) {
+              console.log('setting the damn state', newState.lineFilter);
```
Comment (matyax) [inline src/Components/ServiceScene/LogsListScene.tsx:184]:
```suggestion
```

**Comment by matyax** [comment]
Stretch goal, but I would love if you can include a minimal validation to the input, so we can tell the user that something is going to produce errors beforehand:

imagen

Definitely as a follow up, but if we can release it along with this it would be perfect.

**Comment by matyax** [comment]
Idea: could we use a query with line limit 1, not sure if instant would help, to let the user know that a line filter will produce no results before applying? (If the interaction gives time)

**Comment by matyax** [comment]
Besides what I commented, this seems ready to be merged.

**Comment by gtk-grafana** [comment]
> The field tab seems to be acting up. Takes me a little bit but I manage to reproduce it to the end. Basically query with error > go to tabs > 0 fields > back to logs > remove error > fields stay in 0.

Good catch, I forgot to subscribe to the line-filters and update the detected_fields on change. Will have a fix shortly.

**Review by matyax — APPROVED** [review]
👏 👏
